### DISCUSSION TITLE
Can ALPypeRL affect model logic in a process flowchart?

### DISCUSSION THREAD

**Discussion by cosmin-aron** [MarcEscandell/ALPypeRL#7] [Q&A] [answered]
Hello,
I've been trying to work with ALPypeRL in a more complex model with multiple process flowcharts that run simultaneously and am having problems understanding how ALPypeRL is controlling the model in this case. 

I have everything set up in terms of MDP in the model and in theory it should run properly. Say we have 2 processes with the following blocks, and the last block in Process 1 frees a specific agent from Wait1 in Process 2. When that agent enters the Wait2 block, the `requestAction()` method is called. Then at the end of the `takeAction()` method, that agent is freed from the Wait2 block. 

```mermaid
---
title: Process 1
---
flowchart LR
    A((Source)) --> B[Wait] --> C[Seize] --> 
    D["Free agent from 
    Wait1 in Process 2"]
style D fill:#7f7d9c

```
```mermaid
---
title: Process 2
---
flowchart LR
    A((Source)) -->B[Wait1] --> C[Wait2] --> 
    D[Seize]
style B fill:#7f7d9c
style C fill:#dd9499
```

Of course, this diagram is just a simplification. As far as I know, the `free()` method on a Wait block schedules an event internally, which will send an agent out of that block.

My question is can ALPypeRL interfere with how the agents are freed from Wait blocks? Does it simply pause and resume the model engine when `requestAction()` is called, or does it also step the model? I'm trying to understand if it can work with this approach. I suspect that this can become problematic if multiple agents enter in Wait2 where `requestAction()` is called.

Does anyone know if this is indeed the case here or if the problem lies somewhere else?

**Comment by MarcEscandell** [comment, marked as answer]
Hi @cosmin-aron! Thanks for sharing your question here. I think I understand the problem you are trying to solve, but I do not really grasp the issue you are facing.

In any case, let me share how events unfold in _ALPypeRL_ (I hope it can help clarify your questions):

1. `requestAction()` is called from your simulation side.
2. The python side receives the request and generates a new action.
3. The new action is passed to your simulation and consumed by the `takeAction(RLAction action)` function.

All of the above happens at `t=t`. Or, in other words, there's no gap in between. This is instantaneous. 

In your case, it seems that the problem might come from AnyLogic directly, not _ALPypeRL_. I don't think you can free an agent (let it enter `Wait2` block) and then, at the same time, free it from the same block. Is it the same agent you let in and then free? Have you tried to run your example without _ALPypeRL_? Say, write a script that does exactly that, lets the agent enter `Wait2` and then be freed right after.

As you rightly point out, when you free an agent by calling `free`, AnyLogic internally schedules an event. However, the `requestAction()` and `takeAction` happen one after the other. This means that by the time `takeAction` is called, your agent might not have yet entered the block, so it cannot be freed.

Sorry if my answer is a bit cumbersome. I'm just making my own assumptions. Let me know if I helped.

**Reply by cosmin-aron** [reply]
Hi @MarcEscandell, thank you for the detailed explanation and I apologise if the description was a bit confusing. What I meant is that when an agent enters `Wait2`, the `requestAction()` is called (i.e., in the _On enter_ block), then at the end of `requestAction()` I call `wait2.free(agent)`, which should work fine. 

There were some issues in the model which prevented `requestAction()` to execute (because no agents were entering the block). But you have addressed my concern regarding how the events unfold in ALPypeRL. If `requestAction()` and `takeAction()` happen without stepping the model, then it shouldn't affect how agents are freed from process blocks. That said, I suppose it should also work if multiple agents enter the block at the same time. Normally it should request an action for each of them.
